x-pack/qa/rolling-upgrade-basic/build.gradle

```diff
@@ -1,126 +1,113 @@
             setting 'xpack.watcher.enabled', 'false'
             setting 'xpack.license.self_generated.type', 'basic'
-            setting 'node.name', "upgraded-node-${stopNode}"
+            javaHome = project.file(project.ext.runtimeJavaHome)
         }
     }
 
-    Task oneThirdUpgradedTest = tasks.create(name: "${baseName}#oneThirdUpgradedTest", type: RestIntegTestTask)
+    tasks.register("${baseName}#oldClusterTest", RestTestRunnerTask) {
+        useCluster testClusters."${baseName}"
+        mustRunAfter(precommit)
+        systemProperty 'tests.rest.suite', 'old_cluster'
+        systemProperty 'tests.upgrade_from_version', version.toString().replace('-SNAPSHOT', '')
+        nonInputProperties.systemProperty('tests.rest.cluster', "${-> testClusters."${baseName}".allHttpSocketURI.join(",") }")
+        nonInputProperties.systemProperty('tests.clustername', "${-> testClusters."${baseName}".getName() }")
+    }
 
-    configureUpgradeCluster("oneThirdUpgradedTestCluster", oldClusterTestRunner, 0,
-        // Use all running nodes as seed nodes so there is no race between pinging and the tests
-        { [oldClusterTest.nodes.get(1).transportUri(), oldClusterTest.nodes.get(2).transportUri()] })
 
-    Task oneThirdUpgradedTestRunner = tasks.getByName("${baseName}#oneThirdUpgradedTestRunner")
-    oneThirdUpgradedTestRunner.configure {
+    tasks.register("${baseName}#oneThirdUpgradedTest", RestTestRunnerTask) {
+        dependsOn "${baseName}#oldClusterTest"
+        useCluster testClusters."${baseName}"
+        doFirst {
+            testClusters."${baseName}".nextNodeToNextVersion()
+        }
+        nonInputProperties.systemProperty('tests.rest.cluster', "${-> testClusters."${baseName}".allHttpSocketURI.join(",")}")
+        nonInputProperties.systemProperty('tests.clustername', "${-> testClusters."${baseName}".getName()}")
         systemProperty 'tests.rest.suite', 'mixed_cluster'
         systemProperty 'tests.first_round', 'true'
-        finalizedBy "${baseName}#oldClusterTestCluster#node1.stop"
+        systemProperty 'tests.upgrade_from_version', bwcVersion.toString().replace('-SNAPSHOT', '')
     }
 
-    Task twoThirdsUpgradedTest = tasks.create(name: "${baseName}#twoThirdsUpgradedTest", type: RestIntegTestTask)
-
-    configureUpgradeCluster("twoThirdsUpgradedTestCluster", oneThirdUpgradedTestRunner, 1,
-        // Use all running nodes as seed nodes so there is no race between pinging and the tests
-        { [oldClusterTest.nodes.get(2).transportUri(), oneThirdUpgradedTest.nodes.get(0).transportUri()] })
-
-    Task twoThirdsUpgradedTestRunner = tasks.getByName("${baseName}#twoThirdsUpgradedTestRunner")
-    twoThirdsUpgradedTestRunner.configure {
+    tasks.register("${baseName}#twoThirdsUpgradedTest", RestTestRunnerTask) {
+        dependsOn "${baseName}#oneThirdUpgradedTest"
+        useCluster testClusters."${baseName}"
+        doFirst {
+            testClusters."${baseName}".nextNodeToNextVersion()
```

Comment:
I do tend to agree with Ryan's concerns that this sort of "mutable node" stuff makes it a bit hard to follow. To determine what the state of the cluster is for any given task you have to manually follow the task graph backwards, which is not an easy thing to do.

As we discussed before though, I think this is fine so long as we agree this is a temporary situation until we develop a standard high level model for external cluster testing that supports all the current use cases.

Reply:
I agree, we will need a model.